### PR TITLE
EXAMPLE : Interpreter grounded Neural Program Synthesis [WIP]

### PR DESCRIPTION
**Pull request by reshinthadithyan**
This PR adds an example of using `trlx` to ground LMs into the interpreter with a toy list manipulation DSL to make them generate more coherent programs to satisfy IO conditions.
### Questions:
1. I want to do extensive ablation studies on the impact of grounding Models with Interpreter. So would need a folder to add plots and `ReadME.md`. Does the current directory structure look good ?

# Interpreter Grounded Program Synthesis
*Program synthesis* is the task of automatically generating programs that solve a given task by satisfying an IO condition. In Neural Program Synthesis the synthesizer is a neural network which is a Language Model that takes in an input/output pair and tries to generate the program in the defined toy DSL's Grammar.

## Toy List Manipulation DSL Grammar
The DSL has the following grammar:
```
list_expr := list[int]
integer := -5 | -4 | -3 | -2 | -1 | 0 | 1 | 2 | 3 | 4 | 5
statement :=
          | take(list_expr,integer)
          | drop(list_expr,integer)
          | reverse(list_expr)
          | sort_asc(list_expr)
          | sort_des(list_expr)
          | add_n(list_expr,integer)
          | sub_n(list_expr,integer)
          | mul_n(list_expr,integer)
          | expand_copy(list_expr)


```
To generate training/testing data run, `python3 -m lang`. The dataset would be saved in `./dataset/train.json` and `./dataset/test.json`. To use the processed dataset refer to this [google drive link](https://drive.google.com/drive/folders/1093FlJA0MF7gh25yi4-__yU6Fj-onK1v?usp=share_link).    
Each datapoint in the dataset would look like,
```json
    {"input": "Input: [4, -2, 0, 0, 5, 5] Output: [25, 25, 20, 0, 0, -10] Function:",
    "output": "sort_des(reverse(mul_n(sort_asc(sort_asc([4, -2, 0, 0, 5, 5])),5)))"}
```
## Training with TRLX
Run `python3 -m train_trlx.py` to run the training with grounded interpreter. The `reward_fn`, would return `-1` if a sample generated is of invalid syntax. it would return `0.5` if the generated syntax is valid but doesn't satisfy IO condition.

### CODE REVIEW

File: examples/experiments/grounded_program_synthesis/lang.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+def init_random_input(len_range:int=5,value_gen=5)->list:
+    len_gen = random.randint(2, len_range+1)
+    value_range = list(range(-value_gen, value_gen+1))
+    output = []
+    for index in range(len_gen):
+        value_gen = random.choice(value_range)
+        output.append(value_gen)
+    return output
+
+
```
Comment (cat-state) [inline examples/experiments/grounded_program_synthesis/lang.py:17]: Could you comment this file more? Maybe to separate the sections and an explanation at the top for the structure of the language/interpreter

File: examples/experiments/grounded_program_synthesis/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Interpreter Grounded Program Synthesis
+*Program synthesis* is the task of automatically generating programs that solve a given task by satisfying an IO condition. In Neural Program Synthesis the synthesizer is a neural network which is a Language Model that takes in an input/output pair and tries to generate the program in the defined toy DSL's Grammar.
+
+## Toy List Manipulation DSL Grammar
+The DSL has the following grammar:
```
Comment (cat-state) [inline examples/experiments/grounded_program_synthesis/README.md:5]: In addition to the DSL grammar, add some snippets

**Review by cat-state — COMMENTED** [review]
thanks for this example @reshinthadithyan ! it mostly looks good aside from a few things. Could you also format the python files with black?

**Comment by LouisCastricato** [comment]
This looks fantastic! I think we should upload a trained model to hugging face hub that goes along with this tutorial, and allow people to validate their model against our uploaded one. What do you think Reshinth?

File: examples/experiments/grounded_program_synthesis/lang.py (hunk 1 above)
Comment (LouisCastricato) [reply examples/experiments/grounded_program_synthesis/lang.py:17]: Strong agree. The documentation here is not clear.

Comment (reshinthadithyan) [reply examples/experiments/grounded_program_synthesis/lang.py:17]: Adding more documentation, thanks for the review.

**Comment by reshinthadithyan** [comment]
@LouisCastricato Yes, the idea is to give a fully reproducible script for the anyone to train it by themselves and validate.

**Comment by LouisCastricato** [comment]
LGTM, ready to merge with your approval @cat-state

File: examples/experiments/grounded_program_synthesis/lang.py (hunk 1 above)
Comment (reshinthadithyan) [reply examples/experiments/grounded_program_synthesis/lang.py:18]: Added more documentation on the sampler.

File: examples/experiments/grounded_program_synthesis/README.md (hunk 2 above)
Comment (reshinthadithyan) [reply examples/experiments/grounded_program_synthesis/README.md:5]: Added example snippets to showcase the atomic functions.

**Comment by cat-state** [comment]
LGTM! thanks @reshinthadithyan !

**Review by cat-state — APPROVED** [review]

**Comment by LouisCastricato** [comment]
Precommit is failing. Will merge after resolution.

**Review by LouisCastricato — APPROVED** [review]

**Pull request merged**